### PR TITLE
Add fmt::Formatter::pad_integral_with.

### PR DESCRIPTION
**Pull request by clarfonthey**
Because properly formatting integers can be a bit of a chore, the `pad_integral` method of `fmt::Formatter` is very convenient. However, this method doesn't work very well with bignums, because it requires allocating a string on the heap before writing.

However, `pad_integral_with` will allow making this a bit easier, leveraging the existing implementation. This will, instead of requiring a string, take in a closure which writes out the "string" and a length of the "string." This means that writing a bignum, for example, simply requires computing the number of digits for the string (which may be already computed) and then providing a closure to write the digits.

This should make following the formatting conventions for large integers much easier, although I understand if this form of API isn't ideal.

### CODE REVIEW

File: src/libcore/fmt/mod.rs [hunk 1]
```diff
@@ -1037,12 +1037,31 @@ impl<'a> Formatter<'a> {
...
+        is_nonnegative: bool,
+        prefix: &str,
+        buf: &str
+    ) -> Result {
+        self.pad_integral_with(is_nonnegative, prefix, buf.len(), |f| f.write_str(buf))
+    }
+
+    /// Equivalent to `pad_integral`, allowing a custom closure for writing the integer.
+    ///
+    /// This changes the `buf` argument to two arguments, `write_digits` and `num_digits`.
```
Comment (kennytm) [inline src/libcore/fmt/mod.rs]: `num_digits` → `len_digits` ?

**Comment by shepmaster** [comment]
Ping @rust-lang/libs — haven't heard from @alexcrichton in 7 days. Reassigning to ....

r? @aturon

**Comment by alexcrichton** [comment]
Sorry about that, been traveling!

Thanks for the PR @clarcharr, although I'm personally in favor of helpers like this existing on crates.io rather than in the standard library itself. These exist today mostly for the primitive implementations in libcore itself, which if we were to stablize afresh today probably wouldn't even expose it...

**Comment by clarcharr** [comment]
That makes sense! In that case it may make sense to eventually deprecate these functions in the long term.

**Comment by alexcrichton** [comment]
If you agree, should we close this?

**Comment by clarcharr** [comment]
@alexcrichton yup! I'll close this for now and add a PR to deprecate the original function later.

**Pull request closed without merging**